Journal concept

daisieh opened the pull request.

daisieh committed:
Consolidating PublicationBean into Manuscript

daisieh committed:
remove context caching

daisieh committed:
Add convenience methods for add/clear metadata

daisieh committed:
DryadJournalConcept wraps journal-specific concepts

daisieh committed:
Tests

daisieh committed:
Update multiple concepts at a time

daisieh committed:
can query Organizations by either ISSN or by code

daisieh commented:
This is deployed on dev for testing.

Testing for concepts:
- Start a submission with a new journal:
  Start with 'manuscript is published' and create a new journal.
- Find the new journal via status "CANDIDATE"
- Update the new journal to "ACCEPTED" via API
- See that the new journal concept has the updated status
- Get the metadata for the journal concept from the API
- Add some journal metadata values to the concept and update via the API
  - set integrated and allowsReview to true
- Create a submission using the newly integrated concept
  - see that the new journal comes up in the review journal pulldown
- Check the old submission: it should be integrated now as well
- Check that a new concept that has a payment plan actually behaves as such when using it for a submission
- GET several concepts through the API
  - update those concepts by changing some metadata and pushing back through the API
  - verify the changes through the GUI
- GET concepts and manuscripts through the API using ISSNs instead of journal codes
- Trying to POST to create new concepts that have the same name as ones that already exist should fail
- Trying to PUT to update concepts that don't exist should fail

NOTE: Payment Plan metadata overrides subscriptionPaid

daisieh committed:
return null if an object already exists before creation

daisieh committed:
check for object's existence directly inside create/update methods

daisieh committed:
Only use CF_ACCEPTED if the journal concept is accepted, not candidate.

daisieh committed:
unset

daisieh committed:
Add an accessor to get journal concept by ISSN

rnathanday commented on dspace/modules/api/src/main/java/org/datadryad/rest/storage/rdbms/OrganizationDatabaseStorageImpl.java:
```diff
@@ -193,8 +216,24 @@ protected void deleteObject(StoragePath path) throws StorageException {
+            context.restoreAuthSystemState();
+            completeContext(context);
+        } catch (Exception ex) {
+            throw new StorageException("Exception updating organization: " + ex.getMessage(), ex);
```
missing context.abort() in this catch block

daisieh replied:
the method completeContext has the abort in it.

daisieh replied:
oh but yeah, I can see that it could have thrown a different exception? I'll add it in anyway.

rnathanday commented on dspace/modules/api/src/main/java/org/datadryad/rest/storage/rdbms/OrganizationDatabaseStorageImpl.java:
```diff
@@ -170,21 +165,49 @@ protected void addResults(StoragePath path, List<Organization> organizations, St
-        } catch (SQLException ex) {
-            throw new StorageException("Exception reading organization", ex);
+        } catch (Exception e) {
+            throw new StorageException("Exception reading organization: " + e.getMessage());
```
missing context.abort() in this catch block

daisieh replied:
the method completeContext has the abort in it.

rnathanday commented on dspace-api/src/main/java/org/dspace/authority/AuthorityValue.java:
```diff
@@ -468,16 +468,14 @@ public void updateConceptFromAuthorityValue(Concept concept) throws SQLException
                 {
                     // Add it to the list
-                    concept.addMetadata(schema,element,qualifier,"",value,null,-1);
+                    concept.addMetadata(context, schema,element,qualifier,"",value,null,-1);
```
The -1 for confidence seems like it should either be a constant or the default value in the schema

daisieh replied:
Agreed. I don't know why it was put in there like that in the first place. -1 is equal to `Choices.CF_UNSET`.

daisieh committed:
more robust abortContexts

daisieh committed:
use constant instead of hard-coded -1

rnathanday commented on dspace/modules/api/src/main/java/org/dspace/JournalUtils.java:
```diff
@@ -560,240 +349,43 @@ public static void writeManuscriptToDB(Context context, Manuscript manuscript) t
+                // if nothing, use the deprecated metadata files (this can be deleted after we completely move away from them):
+                File file = new File(journalConcept.getMetadataDir() + File.separator + manuscriptNumber + ".xml");
+                try {
+                    result = ManuscriptToLegacyXMLConverter.convertInternalXMLToManuscript(file);
```
convertInternalXMLToManuscript above does not throw an exception (returns null instead)

rnathanday commented on dspace/modules/api/src/main/java/org/datadryad/rest/converters/ManuscriptToLegacyXMLConverter.java:
```diff
@@ -35,4 +37,15 @@ public static void convertToInternalXML(Manuscript manuscript, OutputStream outp
         marshaller.marshal(legacyManuscript, outputStream);
     }
+
+    public static Manuscript convertInternalXMLToManuscript(File file) {
```
call below expects this to throw JAXBException (or similar) like method above

rnathanday commented on dspace/modules/api/src/main/java/org/datadryad/rest/storage/rdbms/ManuscriptDatabaseStorageImpl.java:
```diff
@@ -204,13 +205,35 @@ private static Manuscript getManuscriptById(Context context, String msid, String
+    public List<Manuscript> getManuscriptsMatchingPath(StoragePath path, int limit) throws StorageException {
+        List<Manuscript> manuscripts = new ArrayList<Manuscript>();
+        try {
+            Context context = getContext();
```
possible connection leak from this Context below

rnathanday commented on dspace/modules/api/src/main/java/org/datadryad/rest/storage/rdbms/OrganizationDatabaseStorageImpl.java:
```diff
@@ -193,8 +225,26 @@ protected void deleteObject(StoragePath path) throws StorageException {
+            String name = journalConcept.getFullName();
+            String status = journalConcept.getStatus();
+            context.commit();
+            DryadJournalConcept conceptToUpdate = JournalUtils.getJournalConceptByJournalName(name);
```
This `JournalUtils.getJournalConceptByJournalName` usage is guarded against null pointer exception, but there is one in the file PaymentSystemImpl.java in the method  updateJournal() that is not (sorry -- i can't line comment that file in this diff).

ghost commented:
Code review progress so far:

#### Code Review Checklist
- [ ] Read through the code to ensure it is:
  - [x] Legible
  - [x] Well-commented
  - [x] Free of any "code smells"
  - [ ] Ensure the associated documentation is up to date.
  - [-] Compile it on your local machine and test the new functionality.
  - [ ] Verify that the change is accompanied by adequate testing code. Verify that the code passes the tests.

Integration testing is being done on the dev machine by Debra and Erin, so I am skipping that here.

daisieh committed:
fix line diffs in code review

daisieh commented:
Documentation is updated here: http://wiki.datadryad.org/Journal_Authority_Control_Technology

dfagandryad commented:
```
• successfully edited json file and put changes in concept (changed value of ISSN)
• successfully used POST to add a new concept
• successfully added field with empty value - “”
• is there a way to delete a field? (I deleted field in json file and put, but the field did not delete)
• try to POST an already existing concept - no indication that it already exists
• did POST of new concept with journalID of existing journal - no indication already exists; creates new journal with same journalID
• When concepts have identical journalID, only see the last one added when doing http://dev.datadryad.org/api/v1/organizations/GOT?access_token=1e1d80cc526beef6a25b7a7ee49c535e, but all journals with the same journalID show up when all concepts are viewed
• noted that if fields are not specified, the api creates them with default values
• created json file named blah.son and used that to create a new concept (this means it will be easier to create a script to create new multiple concepts)
• Updated the ISSN in a json file and did a PUT for a concept as follows:
```

```
[dfdryad@dryaddev2 conceptTestingnano blah.json
[dfdryad@dryaddev2 conceptTesting]$ curl --insecure --header "Content-Type:application/json" --request PUT --data-binary @blah.json https://dev.datadryad.org/api/v1/organizations/?access_token=1e1d80cc526beef6a25b7a7ee49c535e
[ {
  "conceptID" : 471,
  "fullName" : "Game of Thorns Journal",
  "description" : "",
  "status" : "ACCEPTED",
  "journalID" : "GOTH",
  "canonicalManuscriptNumberPattern" : "",
  "sponsorName" : "Westerosi Society of America",
  "parsingScheme" : "manuscriptCentral",
  "metadataDir" : "/opt/dryad/submission/journalMetadata/GOTH",
  "issn" : "0093-9999",
  "customerID" : "",
  "memberName" : "",
  "emailsToNotifyOnArchive" : [ ],
  "emailsToNotifyOnReview" : [ ],
  "emailsToNotifyWeekly" : [ ],
  "allowReviewWorkflow" : false,
  "allowEmbargo" : false,
  "integrated" : false,
  "publicationBlackout" : false,
  "paymentPlan" : ""
```

But the GUI for Add new metadata looks like the following:

Metadata
Remove
Name
Value
Language

journal. fullname
Game of Thorns Journal

journal. issn
0093-9999

journal. journalID
GOTH

journal. metadataDir
/opt/dryad/submission/journalMetadata/GOTH

journal. parsingScheme
manuscriptCentral

journal. sponsorName
Westerosi Society of America

daisieh committed:
Clear submit.skipReviewStage metadata before updating.

elizabethdryad commented:
I did some test submissions on Dev.

Payment:
- Sponsored submissions aren't coming through as sponsored ($120 DPC still shows)
- Fee waivers _are_ coming through correctly ($0 balance)
- Didn't test overage charges

If a journal title isn't currently in the system:
- On the "Describe publication" page,  there's a little green bug next to Journal name with mouseover text: "Value has not been confirmed against the controlled vocabulary. Select 'lookup' to search for matching controlled terms." 
  - There is no lookup 
  - this ideally shouldn't be visible to users - if we have to keep the icon, can we remove the text?
  - journal name should still be editable at this point, if possible

If a journal title _is_ currently in the system:
- On the "Describe publication" page,  there's a little green thumb next to Journal name with mouseover text: "This controlled term has been confirmed by the user" 
  - this ideally shouldn't be visible to users - if we have to keep the icon, can we remove the text?

If article status: "in preparation or in review" is selected:
- Submission does not go into review, instead goes directly into curation

If article status "published" is selected:
- If article DOI is provided, article information fills in properly, but Journal name does _not_ have the green thumb on the "Describe publication" page, even if it's already in the system
- If article DOI/PubMedID is not provided and journal name is entered instead, there is no drop-down to assist with entering the correct form of the Journal name
- If article DOI/PubMedID is not provided and journal name is entered instead, and journal name is already in the system, the Journal name _does_ have the green thumb on the "Describe publication" page

ecclary commented:
Testing implementation from GUI (submitter and curator interfaces)
Should test the following

Integrated
-Review, sponsored
-Accepted, sponsored
-Published, sponsored
-Review, non-sponsored
Non-integrated
-Accepted, sponsored
-Accepted, non-sponsored

Test concept file - 1
Published, sponsored
doi:10.5061/dryad.319bj

As submitter:
-Selected published for Molecular Ecology Resources.
-Entered journal name (rather than DOI or PMID)

Upload process seemed fine. 
-Not shown option to select embargo at file upload (as expected)
-I did receive a confirmation message/email of my submission awaiting curator review (as expected)

As curator:
-Submission looks as expected. Journal cover yes. 
-But DPC showing as $120

Take task and click “edit items”
-Journal name appears correctly and green thumbs up
-No email addresses are entered in our metadata. These values are stored at -workflow.review.mailUsers and workflow.archive.mailUsers in curator metadata.
-Check concept file. In concept, these values are stored at journal.notifyOnReview and journal.notifyOnArchive. No values entered for MER. 
-Added 2 email values to MER concept file for journal.notifyOnReview and journal.notifyOnArchive so that we can test if notifications are being sent properly

Test concept file - 2
doi:10.5061/dryad.s1s3d
As submitter:
-Selected published for Molecular Ecology Resources.
-Entered journal name (rather than DOI or PMID)

Upload process seemed fine. 
-Not shown option to select embargo at file upload (as expected)
-I did receive a confirmation message/email of my submission awaiting curator review (as expected)

As curator:
-Submission looks as expected. Journal cover yes. 
-But DPC showing as $120

Take task and click “edit items”
-Journal name appears correctly and green thumbs up
-No email addresses are entered in our metadata. These values are stored at -workflow.review.mailUsers and workflow.archive.mailUsers in curator metadata.
-Check concept file. In concept, these values are stored at journal.notifyOnReview and journal.notifyOnArchive. No values entered for MER. 
-Added 2 email values to MER concept file for journal.notifyOnReview and journal.notifyOnArchive so that we can test if notifications are being sent properly

Test concept file - 3
doi:10.5061/dryad.8880v
As submitter:
-Selected “in review” for Molecular Ecology Resources
-Selected journal name from drop down list
-Did not enter manuscript ID
-Upload process seemed fine. 
-Not showed option to select embargo at file upload (as expected)
-I did receive a confirmation message/email of my submission awaiting curator review (as expected)
-Notice/email of item in review did not go out to my journal.notifyOnReview email addresses (likely because item is awaiting curator review, not in review)
As curator:
-Item is “in curation” rather than in review
-DPC showing as $120 
-Email address are showing up in metadata
-Journal name has green thumbs up
-I look in the concept file and value for “journal.allowReviewWorkflow” is set to true

Take task and click “edit items”
-System prompts me to approve w/o blackout. I do, and my item goes to reauthorize payment.
-No notice of archive is sent (perhaps because item is not in the archive, but rather is in reauthorize payment queue?)
-Claimed item and “skipped payment” option. 
-Data package is now in the (dev) archive
-Notifications of items approved for the archive did go out.

daisieh commented:
@elizabethdryad and @ecclary had problems with submissions showing up as non-sponsored even when journals are supposed to sponsor...

> Payment:
> Sponsored submissions aren't coming through as sponsored ($120 DPC still shows)
> Fee waivers are coming through correctly ($0 balance)
> Didn't test overage charges

After poking around on this, I found that the reason is because we're going to deprecate the journal.subscriptionPaid boolean in favor of journal.paymentPlanType. On dev (and currently on prod), a lot of sponsored journals only have the subscriptionPaid flag set and no value set for the other. We need to update all of the sponsored concepts to have the correct value before we deploy, and @dfagandryad is on that.

@elizabethdryad's feedback:

> On the "Describe publication" page, there's a little green bug next to Journal name with mouseover text: 
> "Value has not been confirmed against the controlled vocabulary. Select 'lookup' to search for matching controlled terms."
> There is no lookup
> this ideally shouldn't be visible to users - if we have to keep the icon, can we remove the text?
> journal name should still be editable at this point, if possible

Yeah, I hate the default authority behaviors in general. My feeling right now is that there is never a situation in our system where the little green bug is the correct behavior. I'd rather remove it entirely for all "authority unset" values.

> On the "Describe publication" page, there's a little green thumb next to Journal name with mouseover text: "This controlled term has been confirmed by the user"
> this ideally shouldn't be visible to users - if we have to keep the icon, can we remove the text?
> This is the current behavior on the system now. I think this should be part of the larger discussion about what we think the various confidence values for controlled vocabularies mean.
> 
> If article status: "in preparation or in review" is selected:
> Submission does not go into review, instead goes directly into curation

I've fixed this with https://github.com/datadryad/dryad-repo/pull/1112/commits/5c5d9b4889dcebdfa35b2139bff663a7be92ebe2

> If article status "published" is selected:
> 
> If article DOI is provided, article information fills in properly, but Journal name does not have the green thumb on the "Describe publication" page, even if it's already in the system
> If article DOI/PubMedID is not provided and journal name is entered instead, there is no drop-down to assist with entering the correct form of the Journal name
> If article DOI/PubMedID is not provided and journal name is entered instead, and journal name is already in the system, the Journal name does have the green thumb on the "Describe publication" page

I still need to look at this behavior set.

daisieh commented:
@ecclary said: 

> -Not shown option to select embargo at file upload (as expected)

This is because I thought that the default value of `allowEmbargo` was false, when it should actually be true. Oops.

Edited to add that actually, MER's concept on dev has `allowEmbargo : false`, so that's just the way it is. @dfagandryad will need to verify the actual values of all of the current concepts on prod before we absolutely deploy this new system, but it should be pretty easy to do programatically.

daisieh committed:
remove subscriptionPaid

daisieh committed:
add default values for DryadJournalConcept metadata

daisieh committed:
If setting a value to blank, actually write out empty string

daisieh committed:
Update JournalUtils indexes when editing concept metadata through the GUI

daisieh committed:
Dryad doesn't handle unsubmitted manuscripts

daisieh committed:
refactor doProcessing logic

daisieh committed:
merge #1121

daisieh committed:
fix comment

daisieh committed:
more refactoring

daisieh committed:
no icon for cf-unset

daisieh committed:
set journalConcept no matter what

daisieh commented:
I have removed the little green bugs for CF-UNSET.

I also changed the behavior to always write out concept metadata, even for empty strings.

> If article status "published" is selected:
> 
> If article DOI is provided, article information fills in properly, but Journal name does not have the green thumb on the "Describe publication" page, even if it's already in the system
> If article DOI/PubMedID is not provided and journal name is entered instead, there is no drop-down to assist with entering the correct form of the Journal name
> If article DOI/PubMedID is not provided and journal name is entered instead, and journal name is already in the system, the Journal name does have the green thumb on the "Describe publication" page

I did a little refactoring of SelectPublicationStep so that the DOI/Pubmed behavior goes through the journal concepts as well. This should work properly now.

daisieh commented:
@rnathanday, if you could do a quick code review of the commits that are mostly in https://github.com/datadryad/dryad-repo/pull/1112/commits/9e8db833a9944f960881c38cd8c8d84b550f0b04 that would be good. There are other little changes too, but that's the big one.

daisieh commented:
@elizabethdryad, @dfagandryad, @ecclary: if you could run your tests on dev one last time, I think we're good to go.

elizabethdryad commented:
@daisieh I re-checked the behavior for article status = "published":

> If article DOI is provided, article information fills in properly, but Journal name does not have the green thumb on the "Describe publication" page, even if it's already in the system
- still true for test journal Amphibia-Reptilia - no icon at all - I assume this is the correct behavior?
- Molecular Ecology okay (green thumb)

> If article DOI/PubMedID is not provided and journal name is entered instead, there is no drop-down to assist with entering the correct form of the Journal name
- this is still true - not sure if this is something we intend to fix right now

> If article DOI/PubMedID is not provided and journal name is entered instead, and journal name is already in the system, the Journal name does have the green thumb on the "Describe publication" page
- still a problem: now journal name does not come through **at all**

elizabethdryad commented:
@daisieh Payment issues seem to be corrected. Sponsored submissions coming through as sponsored.

rnathanday commented on dspace/modules/api/src/main/java/org/dspace/submit/step/SelectPublicationStep.java:
```diff
@@ -308,14 +151,108 @@ private void addEmailsAndEmbargoSettings(DryadJournalConcept journalConcept, Ite
+        String identifier = request.getParameter("article_doi");
+        if (identifier!=null && !identifier.equals("")) {
+            // normalize and validate the identifier
+            Matcher doiMatcher = Pattern.compile("(doi:)*(.+\\/.+)").matcher(identifier);
```
Should this regex be a bit more picky about what it accepts? Here is a recommended one: "^(doi:)?10.\d{4,9}/[-._;()/:A-Z0-9]+$" , although it may need tweaking for Java (from http://blog.crossref.org/2015/08/doi-regular-expressions.html with the optional "doi:" at the beginning).

daisieh replied:
According to the DOI spec, a DOI just has to have a prefix and a suffix separated by a slash, so even though most are 10.something slash something, I don't know that they have to be. Certainly I don't think there's any harm in asking crossref to look up bad ones.

rnathanday replied:
It could make the submission faster to fail and fail meaningfully if Crossref is unavailable, but yes I agree that the slash is sufficient for determining whether to do a DOI or PMID lookup.

ghost commented:
@daisieh the changes in 9e8db83 and following look fine to me, other than tightening up the doi regex.

daisieh committed:
Check for empty journal name.

daisieh commented:
@elizabethdryad:

> > If article DOI/PubMedID is not provided and journal name is entered instead, there is no drop-down to assist with entering the correct form of the Journal name
> > this is still true - not sure if this is something we intend to fix right now

I checked what happens on production now: apparently there isn't a fill-in or drop-down for this particular box.

> > If article DOI/PubMedID is not provided and journal name is entered instead, and journal name is already in the system, the Journal name does have the green thumb on the "Describe publication" page
> > still a problem: now journal name does not come through at all

I fixed this: apparently there is an empty string for journal name that I wasn't checking for.

This is redeployed on dev now, if you wouldn't mind testing one last time.

elizabethdryad commented:
@daisieh 

> > If article DOI/PubMedID is not provided and journal name is entered instead, there is no drop-down to assist with entering the correct form of the Journal name
> > this is still true - not sure if this is something we intend to fix right now
> 
> I checked what happens on production now: apparently there isn't a fill-in or drop-down for this particular box.

Yeah. I kinda feel like maybe there should be, at some point.

> > If article DOI/PubMedID is not provided and journal name is entered instead, and journal name is already in the system, the Journal name does have the green thumb on the "Describe publication" page
> > still a problem: now journal name does not come through at all
> 
> I fixed this: apparently there is an empty string for journal name that I wasn't checking for. This is redeployed on dev now, if you wouldn't mind testing one last time.

Yes, looks good.

daisieh commented:
@rnathanday, is this good to merge? I can do the merge manually if you're happy with the review.

ghost commented:
@daisieh yes, I think it is good to merge.
